classes/local/store/object_file_system.php

```diff
@@ -354,6 +355,28 @@ public function delete_empty_folders($rootpath = null) {
         return $deletedircount;
     }
 
+    /**
+     * Get the name of the files from a specific dir.
+     * @param string $dir
+     * @return array
+     */
+    public function get_filenames_from_dir($dir = '') {
+        if (empty($rootpath)) {
+            $dir = $this->filedir;
+        }
+        $iterator = new RecursiveDirectoryIterator($dir);
```

Comment:
The RecursiveDirectoryIterator scares me a little because I can't see how it works. Is it breadth first or depth first? When does it stop? What order will it process things in?

Lets assume we have a fresh moodle instance from a new client, and there is 10's of millions of local files here because we haven't synced anything to the remote store yet. This task will be trying to grab a full complete recursive listing and stick it all in memory, and then sort it all in memory, and then do something with it.

This all seems like a scaling issue waiting to happen, and testing this on a current prod site which has already been mostly migrated isn't a valid test.

Unless we can guarantee the way the RecursiveDirectoryIterator works, I'd prefer to see an approach with manual recursion so we have full control over how it all works. ie I'd get a root dir listing (max 256 items), then sort it, then drill into the next dir, get a listing, then sort it, then process files 1 by one. Then if we get to a max execution time we mark the file we are at and stop, and at no point did we ever attempt to get a full listing. The next run we would pick up from where we left off.

Reply:
Adding the flags to this like the other one should be perfect